### PR TITLE
Feature/wa 417 token input

### PR DESCRIPTION
**Pull request by sche**
- Add TokenInput component with BigInt arithmetics.
- Unit tests.
- Dependencies.

TODO: Docs & UI tests

### CODE REVIEW

File: SafeUIKit/SafeUIKit/TokenInput.swift [hunk 1]
```diff
@@ -15,11 +15,57 @@ public class TokenInput: UIView {
...
+        let str = String(value)
+        if str.count <= decimals {
+            integerPartTextField.text = ""
+            fractionalPartTextField.text = normalizedFractionalStringForUI(
+                String(repeating: "0", count: decimals - str.count) + str)
+        } else {
+            integerPartTextField.text = String(str.prefix(str.count - decimals)) + String(delimiter)
+            fractionalPartTextField.text = normalizedFractionalStringForUI(String(str.suffix(decimals)))
+        }
+        if decimals == 0 {
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeUIKit/TokenInput.swift:46]: is this a maximum integer digits count? if so, I'd rename variable to `maximumFractionDigits` (see `NumberFormatter` class for Apple's API)

File: SafeUIKit/SafeUIKit/TokenInput.swift [hunk 2]
```diff
@@ -15,11 +15,57 @@ public class TokenInput: UIView {
...
+            fractionalPartTextField.isEnabled = false
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeUIKit/TokenInput.swift:47]: I think, `integerTextField` and `fractionTextField` names would be better - still the same meaning, but more concise.

File: SafeUIKit/SafeUIKit/TokenInput.swift [hunk 3]
```diff
@@ -15,11 +15,57 @@ public class TokenInput: UIView {
...
+        // maximum possible value of token is 2^256 - 1
+        // String(2^256 - 1).count == 78
+        precondition(decimals >= 0 && decimals <= maxDecimals)
+        precondition(value >= 0 && value <= _2_pow_256_minus_1)
+        self.decimals = decimals
+        self.value = value
+        updateUI()
+    }
+
+    private func updateUI() {
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeUIKit/TokenInput.swift:36]: overall, it feels like dealing with String <-> BigInt conversion should be done in a separate class, `BigNumberFormatter` - as a next step.

File: SafeUIKit/SafeUIKit/TokenInput.swift [hunk 4]
```diff
@@ -15,11 +15,57 @@ public class TokenInput: UIView {
...
+            fractionalPartTextField.isEnabled = false
+        } else if decimals == maxDecimals {
+            integerPartTextField.isEnabled = false
+        }
+    }
+
+    private func normalizedFractionalStringForUI(_ initialString: String) -> String {
+        var fractionalStr = initialString
+        while fractionalStr.last == "0" {
+            fractionalStr = String(fractionalStr.dropLast())
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeUIKit/TokenInput.swift:56]: you could just `fractionalStr.removeLast()`

or 

```
let nonZeroDigits = CharacterSet.decimalDigits.subtracting(CharacterSet(charactersIn: "0"))
if let nonZero = str.rangeOfCharacter(from: nonZeroDigits, options: .backwards, range: nil) {
    str.removeSubrange(nonZero.lowerBound..<str.endIndex)
}
```
to just remove all trailing zeroes.

File: SafeUIKit/SafeUIKit/TokenInput.swift [hunk 5]
```diff
@@ -15,11 +15,57 @@ public class TokenInput: UIView {
...
+            fractionalPartTextField.text = normalizedFractionalStringForUI(String(str.suffix(decimals)))
+        }
+        if decimals == 0 {
+            fractionalPartTextField.isEnabled = false
+        } else if decimals == maxDecimals {
+            integerPartTextField.isEnabled = false
+        }
+    }
+
+    private func normalizedFractionalStringForUI(_ initialString: String) -> String {
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeUIKit/TokenInput.swift:53]: maybe this should be a `stringByRemovingTrailingZeroes()`?

File: SafeUIKit/SafeUIKit/TokenInput.swift [hunk 6]
```diff
@@ -15,11 +15,57 @@ public class TokenInput: UIView {
...
+        var fractionalStr = initialString
+        while fractionalStr.last == "0" {
+            fractionalStr = String(fractionalStr.dropLast())
+        }
+        return fractionalStr
+    }
+
+    private func normalizedFractionalStringForValue(_ initialString: String) -> String {
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeUIKit/TokenInput.swift:61]: and this should be `stringByPaddingWithTrailingZeroes(String,Int)`

File: SafeUIKit/SafeUIKit/TokenInput.swift [hunk 7]
```diff
@@ -15,11 +15,57 @@ public class TokenInput: UIView {
...
+        return initialString + String(repeating: "0", count: decimals - initialString.count)
+    }
+
+    private func normalizedIntegerStringForValue(_ initialString: String) -> String {
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeUIKit/TokenInput.swift:65]: `integerPart(of: String)`?

File: SafeUIKit/SafeUIKit/TokenInput.swift [hunk 8]
```diff
@@ -58,21 +104,51 @@ extension TokenInput: UITextFieldDelegate {
     public func textField(_ textField: UITextField,
                           shouldChangeCharactersIn range: NSRange,
                           replacementString string: String) -> Bool {
-        print("Range: \(range)")
-        print("Replacement string: \(string)")
+        guard CharacterSet.decimalDigits.isSuperset(of: CharacterSet(charactersIn: string)) else {
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeUIKit/TokenInput.swift:107]: hmm... I think, a better alternative is

```
guard string.rangeOfCharacter(from: CharacterSet.decimalDigits.inverted) != nil else {
```

File: SafeUIKit/SafeUIKit/TokenInput.swift [hunk 9]
```diff
@@ -58,21 +104,51 @@ extension TokenInput: UITextFieldDelegate {
...
+
+        let currentText = textField.text ?? ""
+        guard let stringRange = Range(range, in: currentText) else { return false }
+        let updatedText = currentText.replacingCharacters(in: stringRange, with: string)
+            .replacingOccurrences(of: String(delimiter), with: "")
+
+        var expectedFullValueString: String
+        if textField.tag == Field.fractional.rawValue {
+            let oldLength = textField.text?.count ?? 0
+            let newLength = oldLength + string.count - range.length
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeUIKit/TokenInput.swift:119]: why not `newLength = updatedText.count`?

File: SafeUIKit/SafeUIKit/TokenInput.swift [hunk 10]
```diff
@@ -58,21 +104,51 @@ extension TokenInput: UITextFieldDelegate {
...
+            return false
+        }
+
+        let currentText = textField.text ?? ""
+        guard let stringRange = Range(range, in: currentText) else { return false }
+        let updatedText = currentText.replacingCharacters(in: stringRange, with: string)
+            .replacingOccurrences(of: String(delimiter), with: "")
+
+        var expectedFullValueString: String
+        if textField.tag == Field.fractional.rawValue {
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeUIKit/TokenInput.swift:117]: I'd split the bodies into methods that would return corresponding string values (integer or fractional) from the string - that would shorten this method's body length

File: SafeUIKit/SafeUIKit/TokenInput.swift [hunk 11]
```diff
@@ -58,21 +104,51 @@ extension TokenInput: UITextFieldDelegate {
...
+                return false
+            }
+            expectedFullValueString = normalizedIntegerStringForValue(integerPartTextField.text ?? "") +
+                normalizedFractionalStringForValue(updatedText)
+        } else { // integer part
+            let fractionalPartValue = fractionalPartTextField.text ?? ""
+            expectedFullValueString = updatedText + normalizedFractionalStringForValue(fractionalPartValue)
+        }
+
+        guard let newExpectedValue = BigInt(expectedFullValueString), newExpectedValue <= _2_pow_256_minus_1 else {
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeUIKit/TokenInput.swift:130]: I'd add that value must be non-negative

File: SafeUIKit/SafeUIKit/TokenInput.swift [hunk 12]
```diff
@@ -58,21 +104,51 @@ extension TokenInput: UITextFieldDelegate {
...
+            normalizedFractionalStringForValue(fractionalPartTextField.text ?? "")
+        guard let newValue = BigInt(expectedFullValueString), !expectedFullValueString.isEmpty else {
+            value = 0
+            return
+        }
+        value = newValue
+
+        let text = textField.text ?? ""
+        if textField.tag == Field.integer.rawValue && !text.isEmpty {
             textField.text! += String(delimiter)
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeUIKit/TokenInput.swift:152]: maybe just make the delimiter a String, not character?

File: SafeUIKit/SafeUIKit/TokenInput.swift (hunk 4 above)
Comment (sche) [reply SafeUIKit/SafeUIKit/TokenInput.swift:56]: do you think it is more readable?

File: SafeUIKit/SafeUIKit/TokenInput.swift (hunk 8 above)
Comment (sche) [reply SafeUIKit/SafeUIKit/TokenInput.swift:107]: why?

File: SafeUIKit/SafeUIKit/TokenInput.swift (hunk 1 above)
Comment (sche) [reply SafeUIKit/SafeUIKit/TokenInput.swift:46]: Here I try to align with ERC20 specification https://theethereum.wiki/w/index.php/ERC20_Token_Standard
Thats why I call it decimals everywhere

File: SafeUIKit/SafeUIKit/TokenInput.swift (hunk 2 above)
Comment (sche) [reply SafeUIKit/SafeUIKit/TokenInput.swift:47]: ok

File: SafeUIKit/SafeUIKit/TokenInput.swift (hunk 5 above)
Comment (sche) [reply SafeUIKit/SafeUIKit/TokenInput.swift:53]: e.g. normalizedIntegerStringForValue does two things: removing leading zeroes and removing trailing decimal separator. I think we can leave name like that.

File: SafeUIKit/SafeUIKit/TokenInput.swift (hunk 4 above)
Comment (sche) [reply SafeUIKit/SafeUIKit/TokenInput.swift:56]: I did with fractionalStr.removeLast

File: SafeUIKit/SafeUIKit/TokenInput.swift (hunk 9 above)
Comment (sche) [reply SafeUIKit/SafeUIKit/TokenInput.swift:119]: ok

File: SafeUIKit/SafeUIKit/TokenInput.swift (hunk 11 above)
Comment (sche) [reply SafeUIKit/SafeUIKit/TokenInput.swift:130]: redundant, but ok

File: SafeUIKit/SafeUIKit/TokenInput.swift (hunk 12 above)
Comment (sche) [reply SafeUIKit/SafeUIKit/TokenInput.swift:152]: ok

**Pull request closed without merging**